src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs

```diff
@@ -318,46 +420,63 @@ private DkmEvaluationResult GetRootResult(DkmClrValue value, DkmClrType declared
+                        flags: value.EvalFlags,
+                        evalFlags: inspectionContext.EvaluationFlags);
+                    GetResultAsync(dataItem, workList, declaredType, inspectionContext, parent: null, completionRoutine: completionRoutine);
                 }
-
-                ReadOnlyCollection<string> formatSpecifiers;
-                var fullName = this.Formatter.TrimAndGetFormatSpecifiers(resultName, out formatSpecifiers);
-                var dataItem = CreateDataItem(
-                    inspectionContext,
-                    resultName,
-                    typeDeclaringMember: null,
-                    declaredType: declaredType.GetLmrType(),
-                    value: value,
-                    parent: null,
-                    expansionFlags: ExpansionFlags.All,
-                    childShouldParenthesize: this.Formatter.NeedsParentheses(fullName),
-                    fullName: fullName,
-                    formatSpecifiers: formatSpecifiers,
-                    category: DkmEvaluationResultCategory.Other,
-                    flags: value.EvalFlags,
-                    evalFlags: inspectionContext.EvaluationFlags);
-                return GetResult(dataItem, value.Type, declaredType, parent: null);
             }
         }
 
-        internal DkmEvaluationResult GetResult(EvalResultDataItem dataItem, DkmClrType runtimeType, DkmClrType declaredType, EvalResultDataItem parent)
+        internal void GetResultAsync(EvalResultDataItem dataItem, DkmWorkList workList, DkmClrType declaredType, DkmInspectionContext inspectionContext, EvalResultDataItem parent, CompletionRoutine<DkmEvaluationResult> completionRoutine)
         {
-            var value = dataItem.Value; // Value may have replaced (specifically, for Nullable<T>).
-
-            string debuggerDisplayName;
-            string debuggerDisplayValue;
-            string debuggerDisplayType;
-            value.GetDebuggerDisplayStrings(out debuggerDisplayName, out debuggerDisplayValue, out debuggerDisplayType);
+            var value = dataItem.Value; // Value may have been replaced (specifically, for Nullable<T>).
+            DebuggerDisplayInfo displayInfo;
+            if (value.TryGetDebuggerDisplayInfo(out displayInfo))
+            {
+                var targetType = displayInfo.TargetType;
+                var attribute = displayInfo.Attribute;
+                value.EvaluateDebuggerDisplayStringAsync(workList, inspectionContext, targetType, attribute.Name,
+                    displayName => value.EvaluateDebuggerDisplayStringAsync(workList, inspectionContext, targetType, attribute.Value,
+                        displayValue => value.EvaluateDebuggerDisplayStringAsync(workList, inspectionContext, targetType, attribute.TypeName,
+                            displayType => completionRoutine(GetResult(inspectionContext, dataItem, declaredType, displayName, displayValue, displayType, parent)))));
+            }
```

Comment:
What I *hope* will happen is that it bubbles back up -> GetRootResultAndContinue -> IDkmClrResultProvider.GetResult, where it will be caught there and then the outermost 'completionRoutine' is invoked.  Does that seem right?